### PR TITLE
Change newrelic-log-ingestion install mechanism

### PR DESCRIPTION
**Pull request by MattWhelan**
Instead of a stack that refers to a SAR app, just install
the SAR app. This removes the need for CAPABILITY_AUTO_EXPAND.

The migration from nested to unnested is complicated. I learned a lot about CloudFormation :) 

I also overhauled how the upgrade CLI works: we now treat an absent CLI option as "Preserve the current value" rather than "Enforce the default during the upgrade". This distinction became important while developing the migration scenario, and I think the new behavior will be less surprising to the user.

### CODE REVIEW

File: newrelic_lambda_cli/integrations.py [hunk 1]
```diff
@@ -77,74 +77,64 @@ def create_role(session, role_policy, nr_account_id):
...
+    capabilities = ["CAPABILITY_IAM"]
+
+    if role_name is not None:
+        parameters.append({
+            "Name": "FunctionRole",
+            "Value": role_name
+        })
+        capabilities = []
+
+    change_set = client.create_cloud_formantion_change_set(
```
Comment (kolanos) [inline newrelic_lambda_cli/integrations.py]: s/create_cloud_formantion_change_set/create_cloud_formation_change_set

File: newrelic_lambda_cli/integrations.py [hunk 2]
```diff
@@ -77,74 +77,64 @@ def create_role(session, role_policy, nr_account_id):
 
 
 def create_log_ingestion_function(
-    session, nr_license_key, enable_logs=False, memory_size=128, timeout=30
+    session, nr_license_key, enable_logs=False, memory_size=128, timeout=30, role_name=None
 ):
-    client = session.client("cloudformation")
+    client = session.client("serverlessrepo")
```
Comment (kolanos) [inline newrelic_lambda_cli/integrations.py]: I think we want to remove the SAR dependency.

File: newrelic_lambda_cli/integrations.py [hunk 3]
```diff
@@ -77,74 +77,64 @@ def create_role(session, role_policy, nr_account_id):
...
-                {"ParameterKey": "Timeout", "ParameterValue": str(timeout)},
-            ],
-            Capabilities=["CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"],
-        )
-        click.echo(
-            "Waiting for stack update to complete, this may take a minute... ", nl=False
-        )
-        client.get_waiter("stack_update_complete").wait(StackName=stack_name)
-        success("Done")
+    # TODO: we need to detect an old-style nested install and unwrap it
```
Comment (kolanos) [inline newrelic_lambda_cli/integrations.py]: I think we should address this before merging this change to ensure a clean update path.

Comment (MattWhelan) [reply newrelic_lambda_cli/integrations.py]: Definitely.

File: newrelic_lambda_cli/cli/integrations.py [hunk 4]
```diff
@@ -163,57 +171,55 @@ def uninstall(aws_profile, aws_region, aws_permissions_check, nr_account_id, for
 @add_options(AWS_OPTIONS)
 @click.option(
     "--enable-logs/--disable-logs",
-    default=False,
+    default=None,
```
Comment (kolanos) [inline newrelic_lambda_cli/cli/integrations.py]: This will also require a README update I think. Think `default=None` can b omitted?

File: newrelic_lambda_cli/cli/integrations.py [hunk 5]
```diff
@@ -163,57 +171,55 @@ def uninstall(aws_profile, aws_region, aws_permissions_check, nr_account_id, for
...
 )
 @click.option(
     "--memory-size",
     "-m",
-    default=128,
+    default=None,
     help="Memory size (in MiB) for the log ingestion function",
     metavar="<size>",
-    show_default=True,
+    show_default=False,
```
Comment (kolanos) [inline newrelic_lambda_cli/cli/integrations.py]: This can also be omitted.

File: newrelic_lambda_cli/cli/integrations.py [hunk 6]
```diff
@@ -163,57 +171,55 @@ def uninstall(aws_profile, aws_region, aws_permissions_check, nr_account_id, for
...
     type=click.INT,
 )
-@add_options(NR_OPTIONS)
 @click.option(
     "--timeout",
     "-t",
-    default=30,
+    default=None,
```
Comment (kolanos) [inline newrelic_lambda_cli/cli/integrations.py]: Why this change?

File: newrelic_lambda_cli/cli/integrations.py [hunk 7]
```diff
@@ -163,57 +171,55 @@ def uninstall(aws_profile, aws_region, aws_permissions_check, nr_account_id, for
...
     help="Timeout (in seconds) for the New Relic log ingestion function",
     metavar="<secs>",
-    show_default=True,
+    show_default=False,
     type=click.INT,
 )
+@click.option(
```
Comment (kolanos) [inline newrelic_lambda_cli/cli/integrations.py]: Needs to be added to README also.

File: newrelic_lambda_cli/cli/integrations.py [hunk 8]
```diff
@@ -163,57 +171,55 @@ def uninstall(aws_profile, aws_region, aws_permissions_check, nr_account_id, for
...
+    "--role-name",
+    default=None,
```
Comment (kolanos) [inline newrelic_lambda_cli/cli/integrations.py]: Believe this can be ommitted.

File: newrelic_lambda_cli/cli/integrations.py (hunk 6 above)
Comment (MattWhelan) [reply newrelic_lambda_cli/cli/integrations.py]: If the CLI has a default value, we can't tell, later on, whether the user specified 30, or just didn't specify a value. For upgrade, an unspecified value now means, "Preserve the existing value". If they specified 120 at install time, then we want to make it easy to preserve that, rather than forcing the user to remember all their settings from the install when they upgrade.

File: newrelic_lambda_cli/cli/integrations.py [hunk 9]
```diff
@@ -58,6 +58,13 @@ def register(group):
     show_default=True,
     type=click.INT,
 )
+@click.option(
```
[the remark this answers is not in the document]
Comment (MattWhelan) [reply newrelic_lambda_cli/cli/integrations.py]: Good point. I'll go through the readme and make appropriate changes. I'll also take a pass through the docs on the docs site.

File: newrelic_lambda_cli/cli/integrations.py [hunk 10]
```diff
@@ -58,6 +58,13 @@ def register(group):
...
+    "--role-name",
+    default=None,
+    help="The name of a pre-created execution role for the log ingest function",
+    metavar="<role_name>",
+    show_default=False,
```
[the remark this answers is not in the document]
Comment (MattWhelan) [reply newrelic_lambda_cli/cli/integrations.py]: Cool, I'll do that.

File: newrelic_lambda_cli/cli/integrations.py (hunk 4 above)
Comment (MattWhelan) [reply newrelic_lambda_cli/cli/integrations.py]: It looks like for boolean flags like this, the default default is False. So we need `default=None` here.

File: newrelic_lambda_cli/integrations.py [hunk 11]
```diff
@@ -76,93 +79,256 @@ def create_role(session, role_policy, nr_account_id):
...
+            "DeletionPolicy"
+        ] = "Retain"
+
+        # We can't change props during import, so let's set them to their current values
+        lambda_client = session.client("lambda")
+        old_props = lambda_client.get_function_configuration(
+            FunctionName="newrelic-log-ingestion"
+        )
+        old_role_name = old_props["Role"].split(":role/")[-1]
+        old_nr_license_key = old_props["Environment"]["Variables"]["LICENSE_KEY"]
```
Comment (kolanos) [inline newrelic_lambda_cli/integrations.py]: Should we preserve any other environment variables the user has set on the log ingestion function? The new template here only has a subset of the environment variables that can be set. I'm wondering if we should expose all the supported environment variables so they don't get clobbered during an update. Thinking specifically about debug mode and endpoint overrides.

Comment (MattWhelan) [reply newrelic_lambda_cli/integrations.py]: This mechanism only works with props. We're using the env vars as a source for prop values, because that works with our existing templates.

In general, there isn't a way to preserve arbitrary env vars across template updates, as far as I can tell.

Comment (MattWhelan) [reply newrelic_lambda_cli/integrations.py]: Testing the upgrade with an older function that has additional env vars, it looks like those vars get preserved. So I think we're OK here.
